### PR TITLE
Convert to compatible NumPy dtype for MaskedArray to_numpy

### LINKED ISSUE #37460: ENH: support casting Int arrays with nulls to np.float?
xref https://github.com/numpy/numpy/issues/17659

> I would expect numpy to automatically convert such arrays into a float type and fill with np.nan. However, for some reason it converts it to object:
> 
> `pd.DataFrame({'col': [1, np.nan, 3]}).astype('UInt8').values.dtype`
> 
> which leads to errors like this one
> 
> `np.nanmax(pd.DataFrame({'col': [1, np.nan, 3]}).astype('UInt8').values)`
> 
> returns "TypeError: boolean value of NA is ambiguous"
> 
> Is that expected ?

Personally I don't have a strong opinion re: whether it's a good idea to support this

### LINKED ISSUE #40630: BUG: series.to_numpy does not work well with pd.Float64Dtype
- [ ] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
from scipy.stats import norm

series = pd.Series([2.0, -2.5, pd.NA], dtype=pd.Float64Dtype())
norm.pdf(series)

Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\XXX-w19Rd76b-py3.9\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1837, in pdf
    cond1 = self._support_mask(x, *args) & (scale > 0)
  File "C:\Users\XXX\AppData\Local\pypoetry\Cache\virtualenvs\XXX-w19Rd76b-py3.9\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 964, in _support_mask
    return (a <= x) & (x <= b)
  File "pandas\_libs\missing.pyx", line 360, in pandas._libs.missing.NAType.__bool__
TypeError: boolean value of NA is ambiguous
```

#### Problem description
My issue is about that pandas does not convert a pd.Series with dtype=pd.Float64Dtype() correctly to a numpy array when I try to use the scipy.stats.norm function. Therefore it cannot handle empty values (pd.NA) and the series needs to be cast right now as a float dtype to properly work which is not ideal. It probably has to do with the fact that the method series.to_numpy is called which does not work well with pd.Float64Dtype.

See also [https://github.com/scipy/scipy/issues/13729]

#### Expected Output
array([0.05399097, 0.0175283 ,        nan])

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.9.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 79 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252
pandas           : 1.2.3
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0
Cython           : None
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.1
sqlalchemy       : 1.4.2
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### LINKED ISSUE #48891: API: should to_numpy() by default return the corresponding type, and raise otherwise?
Currently, `to_numpy()` from a nullable type always returns `dtype` `object`, regardless of whether the Series contains missing values.
E.g.
```python
In [3]: pd.Series([1,2,3], dtype='Int64').to_numpy()
Out[3]: array([1, 2, 3], dtype=object)
```
This isn't great for users/downstream libraries. E.g. https://github.com/matplotlib/matplotlib/issues/23991#issuecomment-1263728565

Users can specify a `dtype` in `to_numpy()`, but if they want to mirror the current behaviour of pandas whereby one has:
```
In [4]: pd.Series([1,2,3], dtype='int64').to_numpy()
Out[4]: array([1, 2, 3])

In [5]: pd.Series([1,2,3], dtype='float64').to_numpy()
Out[5]: array([1., 2., 3.])
```
then they would need to add extra logic to their code.

I'd suggest that instead, `to_numpy()` just always convert to the corresponding `numpy` type. Then:
- if a Series has no `NA`, then `to_numpy()` will just convert to the corresponding `numpy` type:
```
In [4]: pd.Series([1,2,3], dtype='Int64').to_numpy()
Out[4]: array([1, 2, 3])

In [5]: pd.Series([1,2,3], dtype='Float64').to_numpy()
Out[5]: array([1., 2., 3.])
```
- if a Series does have `NA`, then unless it's being converted to `float` (which can handle `nan`), it'll raise:
```
In [8]: pd.Series([1,2,pd.NA], dtype='Int64').to_numpy(int)
---------------------------------------------------------------------------
ValueError: cannot convert to '<class 'int'>'-dtype NumPy array with missing values.
Please either:
- convert to 'float'
- convert to 'object'
- specify an appropriate 'na_value' for this dtype
```

This would achieve:
- better experience for users
- continue to [avoid value-dependent behavior where the metadata (shape, dtype, etc.) depend on the values of the array](https://github.com/pandas-dev/pandas/issues/30038#issuecomment-613427298)

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #48891 (Replace xxxx with the GitHub issue number)
- [ ] closes #40630
- [ ] closes #37460
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This still needs a whatsnew

### CODE REVIEW

**Comment by WillAyd** [comment]
I think this is pretty reasonable. Is this part of a larger set of changes or is this self-contained?

Probably worth a whatsnew

**Comment by phofl** [comment]
This is something we wanted to clean up for a while now, but yes I have to add a whatsnew (didn't get to it yet because this needs more than a single line)

**Comment by phofl** [comment]
Added the whatsnew

**Comment by lukemanley** [comment]
Does this close #48891?

**Comment by phofl** [comment]
yep

**Comment by phofl** [comment]
cc @mroeschke 

I think this is ready for a review

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -1805,6 +1805,8 @@ def map_array(
         return arr.copy()
 
     # we must convert to python types
+    if isinstance(arr.dtype, BaseMaskedDtype):
```
Comment (mroeschke) [inline pandas/core/algorithms.py:1808]: Might be better now if `BaseMaskedArray.map` did something like

```
def map(self, mapper, na_action=None):
    return map_array(self.to_numpy(), mapper, na_action=None)
```

File: pandas/core/methods/to_dict.py [hunk 2]
```diff
@@ -160,7 +171,11 @@ def to_dict(
         return into_c(
             (
                 k,
-                list(map(maybe_box_native, v.to_numpy().tolist()))
+                list(
+                    map(
+                        maybe_box_native, v.to_numpy(na_value=box_na_values[i]).tolist()
```
Comment (mroeschke) [inline pandas/core/methods/to_dict.py:176]: We can't just go to `v.tolist()` here?

File: pandas/io/formats/format.py [hunk 3]
```diff
@@ -1527,6 +1528,8 @@ def _format_strings(self) -> list[str]:
         if isinstance(values, Categorical):
             # Categorical is special for now, so that we can preserve tzinfo
             array = values._internal_get_values()
+        elif isinstance(values, BaseMaskedArray):
+            array = values.to_numpy(na_value=NA)
```
Comment (mroeschke) [inline pandas/io/formats/format.py:1532]: As I discovered in https://github.com/pandas-dev/pandas/pull/56003/files#diff-c2df0eefbdda16cb85d5e0a305db90101f92bfb790a8e989ab2e99b935b7f6b3, maybe we just want to do `to_numpy(dtype=object)` here?

File: pandas/io/parsers/base_parser.py [hunk 4]
```diff
@@ -762,8 +763,15 @@ def _infer_types(
             pa = import_optional_dependency("pyarrow")
             if isinstance(result, np.ndarray):
                 result = ArrowExtensionArray(pa.array(result, from_pandas=True))
+            elif isinstance(result, BaseMaskedArray):
+                if result._mask.all():
+                    result = result.to_numpy(na_value=None)
```
Comment (mroeschke) [inline pandas/io/parsers/base_parser.py:768]: Why do we need this conversion first for the `mask.all()` case?

File: pandas/tests/copy_view/test_array.py [hunk 5]
```diff
@@ -132,21 +132,25 @@ def test_series_array_ea_dtypes(using_copy_on_write):
         assert arr.flags.writeable is True
 
     arr = np.asarray(ser)
-    assert not np.shares_memory(arr, get_array(ser))
-    assert arr.flags.writeable is True
+    assert np.shares_memory(arr, get_array(ser))
```
Comment (mroeschke) [inline pandas/tests/copy_view/test_array.py:135]: Are these related changes?

File: pandas/tests/extension/test_masked.py [hunk 6]
```diff
@@ -163,6 +163,16 @@ def data_for_grouping(dtype):
 
 
 class TestMaskedArrays(base.ExtensionTests):
+    @pytest.mark.parametrize("na_action", [None, "ignore"])
+    def test_map(self, data_missing, na_action):
+        result = data_missing.map(lambda x: x, na_action=na_action)
+        if data_missing.dtype == Float32Dtype():
```
Comment (mroeschke) [inline pandas/tests/extension/test_masked.py:169]: If there a solution to avoid this?

**Review by mroeschke — COMMENTED** [review]
Should we do the same thing for `ArrowExtensionArray.to_numpy`?

File: pandas/core/methods/to_dict.py (hunk 2 above)
Comment (phofl) [reply pandas/core/methods/to_dict.py:176]: No that breaks something but don't remember which test

File: pandas/tests/copy_view/test_array.py (hunk 5 above)
Comment (phofl) [reply pandas/tests/copy_view/test_array.py:135]: Yes, previously we roundtripped through object which triggered a copy, this is a view now! (that's an improvement)

File: pandas/io/parsers/base_parser.py (hunk 4 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:768]: We want a null pyarrow array here, not sure how we can get there otherwise

File: pandas/tests/extension/test_masked.py (hunk 6 above)
Comment (phofl) [reply pandas/tests/extension/test_masked.py:169]: Not with roundtripping through object, maybe your other suggestion above solves this. 

No if your suggestion doesn't work

File: pandas/io/parsers/base_parser.py (hunk 4 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:769]: done

Comment (mroeschke) [reply pandas/io/parsers/base_parser.py:769]: Or maybe it's not even worth doing `to_numpy`, just skipping to `ArrowExtensionArray(pa.array([None]* len(result)))`

**Comment by phofl** [comment]
Can we get this in if ci is green now? Just merged main to resolve whatsnew conflicts

File: doc/source/whatsnew/v2.2.0.rst [hunk 7]
```diff
@@ -105,6 +105,37 @@ For a full list of ADBC drivers and their development status, see the `ADBC Driv
...
+    Out[2]: array([1, 2, 3], dtype=object)
+
+*New behavior:*
+
+.. ipython:: python
+
+    ser = pd.Series([1, 2, 3], dtype="Int64")
+    ser.to_numpy()
+
+The NumPy dtype is determined as follows:
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.0.rst:131]:
```suggestion
The default NumPy dtype (without any arguments) is determined as follows:
```

File: pandas/core/algorithms.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/algorithms.py:1829]: Could we push this into `BaseMaskedArray.map`?

Comment (mroeschke) [reply pandas/core/algorithms.py:1829]: Oh looks like you did this, so this shouldn't be needed anymore?

Comment (phofl) [reply pandas/core/algorithms.py]: Updated but not sure if this will work, I think I tried this already

lets see what CI says

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl
